src/som/interpreter/nodes/ContextualNode.java

```diff
@@ -48,17 +50,35 @@ public final int getContextLevel() {
   }
 
   @ExplodeLoop
-  protected final MaterializedFrame determineContext(final VirtualFrame frame) {
-    SBlock self = (SBlock) SArguments.rcvr(frame);
+  protected final MaterializedFrame determineContext(final Frame frame) {
+    Object next = SArguments.rcvr(frame);
     int i = contextLevel - 1;
 
     while (i > 0) {
-      self = (SBlock) self.getOuterSelf();
+      if (next instanceof SBlock) {
```

Comment:
Graal is going to hate this code. Did you ran the benchmarks to see what's going to happen?